engine/table/src/main/java/io/deephaven/engine/table/impl/sources/regioned/RegionedColumnSourceManager.java

```diff
@@ -768,40 +793,96 @@ public Map<String, Object> getTableAttributes(
+        final List<IncludedTableLocationEntry> tableLocationEntries = includedLocationEntries();
+        final ArrayList<RegionInfoHolder> includedRegions = new ArrayList<>();
+        try (final RowSet.SearchIterator sit = inputRowSet.searchIterator()) {
+            long startSearchFrom = 0;
+            while (includedRegions.size() < maxRegions && sit.advance(startSearchFrom)) {
+                final long regionStartKey = sit.currentValue();
+                final int regionIndex = RegionedColumnSource.getRegionIndex(regionStartKey);
+                if (regionIndex >= tableLocationEntries.size()) {
+                    throw new IllegalStateException("Region index " + regionIndex + " exceeds the number of included " +
+                            "locations: " + tableLocationEntries.size() + " for input row set: " + inputRowSet);
+                }
+                final IncludedTableLocationEntry entry = tableLocationEntries.get(regionIndex);
+                // Based on the cost of computing overlapping row sets, we can push overlap computation into the
+                // parallel region-processing jobs.
+                try (final WritableRowSet overlappingShiftedRowSet = entry.getOverlappingShiftedRowSet(inputRowSet)) {
+                    if (overlappingShiftedRowSet.isEmpty()) {
+                        throw new IllegalStateException(
+                                "Overlapping row set is empty for region index " + regionIndex + " and input row set: "
+                                        + inputRowSet);
+                    }
+                    includedRegions.add(new RegionInfoHolder(entry, overlappingShiftedRowSet.copy()));
+                    // Move to the next region, skipping the current one
+                    final long regionEndKey = getLastRowKey(regionIndex);
+                    startSearchFrom = regionEndKey + 1;
+                }
+            }
+        }
+        return includedRegions;
+    }
+
+    /**
+     * Get all regions and corresponding row set that overlap with the input row set. The user is responsible for
+     * closing the returned row sets.
+     *
+     * @param inputRowSet The input row set
+     * @return A list of included region information
+     */
+    private ArrayList<RegionInfoHolder> getOverlappingRegions(final RowSet inputRowSet) {
+        return getOverlappingRegions(inputRowSet, Integer.MAX_VALUE);
+    }
+
     @Override
     public long estimatePushdownFilterCost(
             final WhereFilter filter,
+            final Map<String, String> renameMap,
             final RowSet selection,
             final RowSet fullSet,
             final boolean usePrev,
             final PushdownFilterContext context) {
-
-        if (includedLocations().isEmpty()) {
-            return 0; // No locations, no cost
+        // We want to test out a small sample of locations to estimate the cost of the filter pushdown, and assume the
+        // rest of the locations are similar.
```

Comment:
This logic now always selects first `n` matching regions instead of trying to sample more broadly. I understand why it was changed and that it would be difficult to restore this functionality.